fs/shm/shmfs_alloc.c

```diff
@@ -0,0 +1,143 @@
+/****************************************************************************
+ * fs/shm/shmfs_alloc.c
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <stdlib.h>
+#include <nuttx/pgalloc.h>
+
+#include "inode/inode.h"
+#include "shm/shmfs.h"
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+FAR struct shmfs_object_s *shmfs_alloc_object(size_t length)
+{
+  FAR struct shmfs_object_s *object;
+  bool allocated = false;
+
+#if defined(CONFIG_BUILD_FLAT)
+  /* in FLAT build, allocate the object metadata and the data in the same
+   * chunk in kernel heap
+   */
+
+  object = kmm_zalloc(sizeof(struct shmfs_object_s) + length);
+  if (object)
+    {
+      object->paddr = (&object->paddr) + 1;
```

Comment:
object->paddr = object->paddr + 1;

Reply:
no, original is correct. in the flat build the actual data is allocated with same malloc call to the end of shmfs object. this takes the address of the last item (paddr) and adds sizeof(void*) to that.

but, I agree that this is confusing. I make it cleaner. maybe

 object->paddr = (FAR char *)object + sizeof(struct shmfs_object_s);

Reply:
sorry, I mean:
object->paddr = object + 1
which is the simplest approach.

Reply:
shortest, but not simplest IMHO. The long version is more descriptive, sizeof(shmfs_object_s) was used above in malloc, so it is nicer to read when you just add that many chars in here.